pkg/ccl/multiregionccl/roundtrips_test.go

```diff
@@ -0,0 +1,135 @@
+// the License. You may obtain a copy of the License at
+//
+//     https://github.com/cockroachdb/cockroach/blob/master/licenses/CCL.txt
+
+package multiregionccl_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/ccl/multiregionccl/multiregionccltestutils"
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvbase"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvcoord"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/sql"
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/stretchr/testify/require"
+)
+
+// TestEnsureLocalReadsOnGlobalTables ensures that all present time reads on
+// GLOBAL tables don't incur a network hop.
+func TestEnsureLocalReadsOnGlobalTables(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// ensureOnlyLocalReads looks at a trace to ensure that reads were served
+	// locally. It returns true if the read was served as a follower read.
+	ensureOnlyLocalReads := func(t *testing.T, rec tracing.Recording) (servedUsingFollowerReads bool) {
+		for _, sp := range rec {
+			if sp.Operation == "dist sender send" {
+				require.True(t, tracing.LogsContainMsg(sp, kvbase.RoutingRequestLocallyMsg),
+					"query was not served locally: %s", rec)
+
+				// Check the child span to find out if the query was served using a
+				// follower read.
+				for _, span := range rec {
+					if span.ParentSpanID == sp.SpanID {
+						if tracing.LogsContainMsg(span, kvbase.FollowerReadServingMsg) {
+							servedUsingFollowerReads = true
+						}
+					}
+				}
+			}
+		}
+		return servedUsingFollowerReads
+	}
+
+	presentTimeRead := `SELECT * FROM t.test WHERE k=2`
+	recCh := make(chan tracing.Recording, 1)
+
+	knobs := base.TestingKnobs{
+		SQLExecutor: &sql.ExecutorTestingKnobs{
+			WithStatementTrace: func(trace tracing.Recording, stmt string) {
+				if stmt == presentTimeRead {
+					recCh <- trace
+				}
+			},
+		},
+	}
+
+	// Create a 3 node cluster. This ensures that every node gets a replica, so
+	// every node should be able to serve present time reads locally.
+	numServers := 3
+	tc, sqlDB, cleanup := multiregionccltestutils.TestingCreateMultiRegionCluster(t, numServers, knobs, nil)
+	defer cleanup()
+
+	_, err := sqlDB.Exec(`CREATE DATABASE t PRIMARY REGION "us-east1" REGIONS "us-east2", "us-east3"`)
+	require.NoError(t, err)
+	_, err = sqlDB.Exec(`CREATE TABLE t.test (k INT PRIMARY KEY) LOCALITY GLOBAL`)
+	require.NoError(t, err)
+
+	errCh := make(chan error)
+	stopWritesChan := make(chan struct{})
+	go func() {
+		i := 0
+		for {
+			select {
+			case <-stopWritesChan:
+				errCh <- nil
+				return
+			case <-time.After(10 * time.Millisecond):
+				_, err = sqlDB.Exec(`INSERT INTO t.test VALUES($1)`, i)
+				i++
+				if err != nil {
+					errCh <- err
```

Comment:
if there is an error here, this blocks until the whole numServer routine completes. is that expected? Should we be using select somewhere below to consume the errCh?